### PR TITLE
Fix automatic state change on LRC GEM-X 5p cDNA PCR XP plate

### PR DESCRIPTION
**Pull request by KatyTaylor**
2 fixes based on UAT testing:

1. Fix automatic state change on LRC GEM-X 5p cDNA PCR XP plate
- It wasn't completing the requests when coming down the SeqOps cDNA Prep route
- It only worked for the Faculty input route
- This is because the only request type specified in the 'work_completion_request_type' config was 'limber_scrna_core_cdna_prep_input'
- Had to enhance the AutomaticStateChange functionality to allow multiple request types

2. De-duplicate requests in the requestsForWell function
- in the aggregation pipeline for scRNA Core, the plates that the submission is done on may contain multiple aliquots
- if this is the case, each aliquot in the well references the same request
- this method was pulling the request from each of the aliquots - presumably with the assumption that each aliquot would have a different request
- in our case, it created duplicates in the array
- there could also theoretically by duplicates if the same request was referenced on the aliquot and in requests_as_source (although I think this shouldn't happen)
- therefore I added de-duplication to the end of the function chain, based on request id

Closes #

#### Changes proposed in this pull request

- 

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: app/javascript/shared/wellHelpers.js [hunk 1]
```diff
@@ -33,8 +33,23 @@ const quadrantTargetFor = function (plateIndex, wellName, rowOffset, colOffset)
   return wellCoordinateToName([destinationColumn, destinationRow])
 }
 
+// Given a well, return the requests that are relevant to it.
 const requestsForWell = function (well) {
-  return [...well.requests_as_source, ...well.aliquots.map((aliquot) => aliquot.request)].filter((request) => request)
+  const arr = [
```
Comment (andrewsparkes) [inline app/javascript/shared/wellHelpers.js:38]: Feels like to me that surely there has to be a more succinct way of doing what's happening in this method(!)
Ask CoPilot? It's good at that.

Comment (andrewsparkes) [reply app/javascript/shared/wellHelpers.js:38]: But don't necessarily trust the first solution it comes up with...

Comment (KatyTaylor) [reply app/javascript/shared/wellHelpers.js:38]:
```
This is what it came up with, which isn't any more succinct, except for removing a lot of the comments :(

If the comments are too much, I can re-write them?

Interesting that it changes `.filter((request) => request)` to `.filter(Boolean); // Filter out falsy values`
I wasn't sure what that was doing, so I left it alone...

function getUniqueRequestsFromWell(well) {
  // Combine requests_as_source and requests from aliquots into one array
  const combinedRequests = [
    ...well.requests_as_source,
    ...well.aliquots.map(aliquot => aliquot.request),
  ].filter(Boolean); // Filter out falsy values

  // Use a Set to remove duplicates based on the request's uuid
  const uniqueRequests = Array.from(new Set(combinedRequests.map(req => req.uuid)))
    .map(uuid => combinedRequests.find(req => req.uuid === uuid));

  return uniqueRequests;
}
```

Comment (andrewsparkes) [reply app/javascript/shared/wellHelpers.js:38]: Nah comments are good.
agree for filter(Boolean) is not clear what that is doing

File: app/javascript/shared/wellHelpers.spec.js [hunk 2]
```diff
@@ -15,4 +15,29 @@ describe('wellHelpers', () => {
...
+      uuid: '3',
+      position: { name: 'A1' },
+      aliquots: [
+        {
+          uuid: '4',
+          request: request1,
+        },
+        {
+          uuid: '5',
+          request: null,
```
Comment (andrewsparkes) [inline app/javascript/shared/wellHelpers.spec.js:33]: not quite following why this is null
is it possible for one aliquot in a well to have a request and the other not?
also, is it possible for request1 to be both in requests_as_source (which it will be if the submission is made on this labware) AND in the aliquot (which is set only on children of the submitted well)?
maybe add some comments describing the scenario?

Comment (andrewsparkes) [reply app/javascript/shared/wellHelpers.spec.js:33]: the real scenario was more that multiple aliquots in the well has the same aliquot.request yes?

File: spec/models/state_changers_spec.rb [hunk 3]
```diff
@@ -125,6 +125,24 @@
           expect(work_completion_creation).to_not have_been_made
         end
       end
+
+      # The ability to have multiple request types in the config was added for scRNA Core pipeline.
+      # The expectation was that any one plate would only have one of the request types on it,
+      # so I haven't tested a plate with a mix of request types.
+      context 'when one of the multiple config request types matches the in progress submissions' do
```
Comment (andrewsparkes) [inline spec/models/state_changers_spec.rb:132]: It occurs to me that aggregation is like cherrypicking. So they may return to the same source and pick a different subset of wells (or even repeat pick some). So multiple requests_as_source in a well is possible. Only one should be active at a time though.
I think this is fine for work completion, as I think the filtering for active requests is happening on the SS side.
But for the de-duplication code further up, do we need to consider filtering requests_as_source for active requests?

Comment (andrewsparkes) [reply spec/models/state_changers_spec.rb:132]: what is the `requests_for_well` method used for again?

File: app/javascript/shared/wellHelpers.spec.js (hunk 2 above)
Comment (KatyTaylor) [reply app/javascript/shared/wellHelpers.spec.js:33]: You're right, it's not representing the real scenario, I was just trying to cover everything in one test (bad practice 😅 ). I'll split them out.

I don't think it's expected for the same request to be in both requests_as_source and on the aliquot... but I wouldn't be surprised if that happens in the future (or in some edge case now), so I thought I should account for it.

Comment (yoldas) [reply app/javascript/shared/wellHelpers.spec.js:33]: requestsForWell filters falsy values out and the test is describing the requestsForWell function. Having a null request covers that case.

File: spec/models/state_changers_spec.rb (hunk 3 above)
Comment (KatyTaylor) [reply spec/models/state_changers_spec.rb:132]: `requestsForWell` is used in `requestsFromPlates` in `requestHelpers.js`

That's then used in `MultiStamp.vue`, where it then calls `requestIsActive` on them.

So `requestsForWell` simply grabs the requests, and they're filtered later.

Comment (yoldas) [reply spec/models/state_changers_spec.rb:132]: requests_as_source implies transfer requests on receptacle I think.

**Review by yoldas — COMMENTED** [review]
Enables multiple work completion request types. 
Extracts unique requests from well. 
Looks good. Waiting until finalising discussions.

**Comment by KatyTaylor** [comment]
Merging to unblock @andrewsparkes , but will come back to improve the js tests
